feat: add support for event archive in EventBridge resource

heitorlessa opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Intro

This PR adds support to create Archives in EventBridge. 

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

`resource_aws_cloudwatch_event_archive - add new resource`


**Tasks**

- [x] Initial CRUD implementation
- [x] Acceptance test
- [x] Update docs
- [x] Investigate answers to open questions

### Open questions as a first time contributor

~~**Q: How do I add support for import?**~~

Update: Fix available in https://github.com/hashicorp/terraform-provider-aws/pull/17270/commits/5ab7d948f9284659fa028719c8f79f610a26735b

~~Tried following other resources that do support but it seems there's something wrongly parametrized:~~

```go
// ...archive.go
	return &schema.Resource{
                ...
		Importer: &schema.ResourceImporter{
			State: schema.ImportStatePassthrough,
		},
```

```go
// ...archive_test.go
....
{
    // also tried without Config key
    Config:            testAccAWSCloudWatchArchiveConfig(archiveName),
    ResourceName:      resourceName,
    ImportState:       true,
    ImportStateVerify: true,
},
```

> Error

```go
    resource_aws_cloudwatch_event_archive_test.go:78: Step 2/2 error running import: 
        Error: Error reading CloudWatch Events archive: InvalidParameter: 1 validation error(s) found.
        - minimum field size of 1, DescribeArchiveInput.ArchiveName.
```

~~**Q: How do I fix acceptance test lint error in CI?**~~

~~Can't see what's wrong by just looking at GH Actions logs. Maybe I missed that in the contrib checklist docs.~~

**A**: Running `terrafmt diff --check --fmtcompat aws/resource_aws_cloudwatch_event_archive_test.go` locally showed the error.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Relates OR Closes #16067

Output from acceptance testing: WIP

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```
$ make fmt && make testacc TESTARGS='-run=TestAccAWSCloudWatchArchive'       
==> Fixing source code with gofmt...
gofmt -s -w ./aws ./awsproviderlint/helper ./awsproviderlint/main.go ./awsproviderlint/passes
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCloudWatchArchive -timeout 120m
=== RUN   TestAccAWSCloudWatchArchive_basic
=== PAUSE TestAccAWSCloudWatchArchive_basic
=== RUN   TestAccAWSCloudWatchArchive_update
=== PAUSE TestAccAWSCloudWatchArchive_update
=== CONT  TestAccAWSCloudWatchArchive_basic
=== CONT  TestAccAWSCloudWatchArchive_update
--- PASS: TestAccAWSCloudWatchArchive_basic (33.35s)
--- PASS: TestAccAWSCloudWatchArchive_update (55.17s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       57.862s

```

heitorlessa commented:
All linting issues fixed - Ready for review

bill-rich commented on aws/resource_aws_cloudwatch_event_archive.go:
```diff
@@ -0,0 +1,200 @@
+
+	log.Printf("[DEBUG] Found Archive: #{*out}")
+
+	// Review Question - Is there a problem in setting more than should name and Arn?
```
I don't see a problem, but to make sure I fully understand, what is your concern?

heitorlessa replied:
Thanks @bill-rich - As it's my first time, I wasn't sure if there was a limit on how many you could set, that's all :)

bill-rich commented on aws/resource_aws_cloudwatch_event_archive.go:
```diff
@@ -0,0 +1,200 @@
+				ForceNew:     true,
+				ValidateFunc: validateCloudWatchEventArchiveName,
+			},
+			"event_source_arn": {
```
This should probably be `ForceNew` since it can't be updated, unless it is expected to be changing on the backend and you don't want to trigger the creation of a new resource.

heitorlessa replied:
Indeed, great catch.

bill-rich commented on aws/resource_aws_cloudwatch_event_archive_test.go:
```diff
@@ -0,0 +1,230 @@
+				Config: testAccAWSCloudWatchArchiveConfig(archiveName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudWatchArchiveExists(resourceName, &v1),
+					resource.TestCheckResourceAttr(resourceName, "name", archiveName),
```
To make sure everything is working as expected, you should check all attributes are set correctly.

bill-rich commented on aws/resource_aws_cloudwatch_event_archive_test.go:
```diff
@@ -0,0 +1,230 @@
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudWatchArchiveExists(resourceName, &v1),
+					resource.TestCheckResourceAttr(resourceName, "retention_days", "7"),
+					testAccCheckResourceAttrEquivalentJSON(resourceName, "event_pattern", "{\"source\":[\"company.team.service\"]}"),
```
You should try updating `description` here also to make sure all the updatable attributes work.

bill-rich commented on website/docs/r/cloudwatch_event_archive.html.markdown:
```diff
@@ -0,0 +1,71 @@
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the new event archive. The archive name cannot exceed 48 chars.
```
Use `characters` instead of the shortened `chars`.

bill-rich reviewed (changes requested):
Looks good! Just a few minor suggestions and questions.

heitorlessa commented:
Addressed all comments: **Changelog entry**, **Test remaining attributes can be updated**, **ForceNew attribute**, and **Doc chars -> characters**. 

CI is failing due to Security Hub test file (non-existent) formatting unrelated to this PR, though

---

```shell
09:36:47 ➜ terraform-aws git:(feat/eventbridge_replay!) go:(🐹 v1.15.7)   find ./aws -type f -name '*_test.go' \
➜     | sort -u \
➜     | grep -v resource_aws_apigatewayv2_domain_name_test.go \
➜     | grep -v resource_aws_ecs_capacity_provider_test.go \
➜     | grep -v resource_aws_efs_file_system_test.go \
➜     | grep -v resource_aws_kinesis_stream_test.go \
➜     | grep -v resource_aws_kms_grant_test.go \
➜     | grep -v resource_aws_quicksight_user_test.go \
➜     | grep -v resource_aws_s3_bucket_object_test.go \
➜     | grep -v resource_aws_sns_platform_application_test.go \
➜     | xargs -I {} terrafmt diff --check --fmtcompat {}

09:39:51 ➜ terraform-aws git:(feat/eventbridge_replay!) go:(🐹 v1.15.7) make fmt && make lint                                                     
==> Fixing source code with gofmt...
gofmt -s -w ./aws ./awsproviderlint/helper ./awsproviderlint/main.go ./awsproviderlint/passes

```

heitorlessa commented on website/docs/r/service_discovery_service.html.markdown:
```diff
@@ -20,7 +20,7 @@ resource "aws_vpc" "example" {
 
 resource "aws_service_discovery_private_dns_namespace" "example" {
-  name        = "example.terraform.local"
+  name        = "example.terraform.internal"
```
this came from merging with `main` after hitting the CI check issue

bill-rich reviewed (approved):
LGTM. Thanks for taking a look at the comments.
